hw/ip/otbn/util/static_check.py

```diff
@@ -0,0 +1,150 @@
+#!/usr/bin/env python3
+# Copyright lowRISC contributors.
+# Licensed under the Apache License, Version 2.0, see LICENSE for details.
+# SPDX-License-Identifier: Apache-2.0
+
+import argparse
+import re
+import struct
+import sys
+from typing import Dict, Iterator, List, Optional, Tuple
+
+from shared.elf import read_elf
+from shared.insn_yaml import Insn, load_insns_yaml
+
+# Load the insns.yml file at module load time.
+try:
+    INSNS_FILE = load_insns_yaml()
+except RuntimeError as err:
+    sys.stderr.write('{}\n'.format(err))
+    sys.exit(1)
+
+
+class OTBNProgram:
+    def __init__(self, symbols: Dict[str, int], insns: Dict[int, int],
+                 data: Dict[int, int]):
+        self.symbols = symbols  # label -> PC
+        self.insns = insns  # PC -> opcode
+        self.data = data  # addr -> data (32b word)
+
+    def _get_pc_at_symbol(self, symbol: str) -> int:
+        return self.symbols[symbol]
+
+    def _get_opcode_at_pc(self, pc: int) -> int:
+        return self.insns[pc]
+
+    def _get_data_at_addr(self, addr: int) -> int:
+        return self.data[addr]
+
+    def _get_insn(self, pc: int) -> Insn:
+        opcode = self._get_opcode_at_pc(pc)
+        mnem = INSNS_FILE.mnem_for_word(opcode)
+        if mnem is None:
+            return IllegalInsn(pc, word, 'No legal decoding')
+
+        return INSNS_FILE.mnemonic_to_insn.get(mnem)
```

Comment:
I'd suggest moving this logic to the constructor, which means we get to memoize the (slow!) decoding. I'd also extract the operands at the same time, so you'd store tuples of an `Insn` and a list of integers. (This is basically how `ProgInsn` works in the RIG).

That way, you don't actually need `_get_insn` or `_get_operands`: it's just an array access.